roofit/roofitcore/src/RooAbsPdf.cxx

```diff
@@ -1160,6 +1163,22 @@ RooAbsReal* RooAbsPdf::createNLL(RooAbsData& data, const RooLinkedList& cmdList)
                                                takeGlobalObservablesFromData).release();
   }
 
+  if (pc.getInt("CodeSquashing")) {
+     std::unique_ptr<RooAbsPdf> pdfClone = RooHelpers::cloneTreeWithSameParameters(*this, data.get());
+     if (addCoefRangeName) {
+        cxcoutI(Fitting) << "RooAbsPdf::fitTo(" << GetName()
+                         << ") fixing interpretation of coefficients of any component to range " << addCoefRangeName
+                         << "\n";
+        pdfClone->fixAddCoefRange(addCoefRangeName, false);
+     }
+
+     return RooFit::ADModeHelpers::translateNLL(
+               std::move(pdfClone), data, createConstr(*pdfClone, /*removeConstraintsFromPdf=*/true),
+               rangeName ? rangeName : "", projDeps, ext, pc.getDouble("IntegrateBins"), doOffset,
+               static_cast<bool>(splitRange), takeGlobalObservablesFromData, pc.getInt("CodePrinting"))
+        .release();
+  }
+
```

Comment:
You have now quite some duplication of code and interfaces with the BatchMode. I think it would be better to call `BatchModeHelpers::createNLL` always here in `RooAbsPdf`, and then pass the object to your translate function. Maybe like this:
```c++
  auto batchMode = static_cast<RooFit::BatchModeOption>(pc.getInt("BatchMode"));
  bool useNewBatchMode = batchMode != RooFit::BatchModeOption::Off && batchMode != RooFit::BatchModeOption::Old;
  bool doCodeSquashing = pc.getInt("CodeSquashing");

  // Code squashing implies new BatchMode
  if(doCodeSquashing) {
    useNewBatchMode = true;
  }

  // Construct BatchModeNLL if requested
  if (useNewBatchMode) {
    std::unique_ptr<RooAbsPdf> pdfClone = RooHelpers::cloneTreeWithSameParameters(*this, data.get());
    if (addCoefRangeName) {
       cxcoutI(Fitting) << "RooAbsPdf::fitTo(" << GetName()
                        << ") fixing interpretation of coefficients of any component to range "
                        << addCoefRangeName << "\n";
       pdfClone->fixAddCoefRange(addCoefRangeName, false);
    }

    std::unique_ptr<RooAbsReal> nll =
           RooFit::BatchModeHelpers::createNLL(std::move(pdfClone),
                                               data,
                                               createConstr(*pdfClone, /*removeConstraintsFromPdf=*/true),
                                               rangeName ? rangeName : "",
                                               projDeps,
                                               ext,
                                               pc.getDouble("IntegrateBins"),
                                               batchMode,
                                               doOffset,
                                               static_cast<bool>(splitRange),
                                               takeGlobalObservablesFromData);

    if (doCodeSquashing) {
       nll = RooFit::ADModeHelpers::translateNLL(std::move(nll), data, pc.getInt("CodePrinting"));
    }

    return nll.release();
  }
```
(replacing the current code branch of the new batch mode)

Like this you can also significantly reduce the interface of your helper:
```c++
std::unique_ptr<RooAbsReal> translateNLL(
                                         std::unique_ptr<RooAbsReal> &&obj,
                                         RooAbsData &data,
                                         bool printCode);
```